REGRESSSION (iOS 17 Beta): Outgoing video and incoming audio do not recover if iOS User receives PSTN/FaceTime call in the middle of the WebRTC call and decline/accepts it from the notification

youennf opened the pull request:
#### 130f20449a4b9edfb76e6b51b93aedecb1bb3f0a
<pre>
REGRESSSION (iOS 17 Beta): Outgoing video and incoming audio do not recover if iOS User receives PSTN/FaceTime call in the middle of the WebRTC call and decline/accepts it from the notification
<a href="https://bugs.webkit.org/show_bug.cgi?id=259884">https://bugs.webkit.org/show_bug.cgi?id=259884</a>
rdar://113534779

Reviewed by Eric Carlson.

When we are interrupted by a phone call, we were sometimes changing the AudioSession Category from PlayAndRecord to Media.
In that case, we might try to activate the AudioSession with Media category and it will succeed given category is Media.
Capture will still fail as setting the category to PlayAndRecord before restarting capture will fail.
In that case, we stay muted but we will not be able to recover properly when getting uninterrupted.

To fix this, we are now sticking to PlayAndRecord when a process has microphone sources which are muted due to interruption.
When we will try to activate the audio session while being interrupted, activation will fail as it is lower priority compared to a phone call,
and we will wait for end of interruption signal.

Covered by added test.

* LayoutTests/fast/mediastream/microphone-interruption-and-audio-session-expected.txt: Added.
* LayoutTests/fast/mediastream/microphone-interruption-and-audio-session.html: Added.
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::wantsToCaptureAudio const):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/platform/audio/PlatformMediaSession.h:
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::countActiveAudioCaptureSources):
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp:
(WebCore::MediaStreamTrackPrivate::interrupted const):
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.h:

Canonical link: <a href="https://commits.webkit.org/269809@main">https://commits.webkit.org/269809@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6f3b886f9a2e6b702d1f67aa7812961b04da12f8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23638 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1754 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24759 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25797 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21821 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23906 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3328 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24174 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22384 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23881 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1301 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20452 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26393 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1090 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21358 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27632 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21556 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21627 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25395 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1066 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18769 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1045 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/21132 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5651 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1479 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1356 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson reviewed (approved):
r=me with the additional comments we discussed

youennf committed:
REGRESSSION (iOS 17 Beta): Outgoing video and incoming audio do not recover if iOS User receives PSTN/FaceTime call in the middle of the WebRTC call and decline/accepts it from the notification
https://bugs.webkit.org/show_bug.cgi?id=259884
rdar://113534779

Reviewed by Eric Carlson.

When we are interrupted by a phone call, we were sometimes changing the AudioSession Category from PlayAndRecord to Media.
In that case, we might try to activate the AudioSession with Media category and it will succeed given category is Media.
Capture will still fail as setting the category to PlayAndRecord before restarting capture will fail.
In that case, we stay muted but we will not be able to recover properly when getting uninterrupted.

To fix this, we are now sticking to PlayAndRecord when a process has microphone sources which are muted due to interruption.
When we will try to activate the audio session while being interrupted, activation will fail as it is lower priority compared to a phone call,
and we will wait for end of interruption signal.

Covered by added test.

* LayoutTests/fast/mediastream/microphone-interruption-and-audio-session-expected.txt: Added.
* LayoutTests/fast/mediastream/microphone-interruption-and-audio-session.html: Added.
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::wantsToCaptureAudio const):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/platform/audio/PlatformMediaSession.h:
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::countActiveAudioCaptureSources):
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp:
(WebCore::MediaStreamTrackPrivate::interrupted const):
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.h:

Canonical link: https://commits.webkit.org/269809@main
